speakerbox/main.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from pathlib import Path
+from typing import Dict, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from datasets import Audio, Dataset, DatasetDict, arrow_dataset, load_metric
+from sklearn.metrics import ConfusionMatrixDisplay
+from transformers import (
+    EvalPrediction,
+    Trainer,
+    TrainingArguments,
+    Wav2Vec2FeatureExtractor,
+    Wav2Vec2ForSequenceClassification,
+    feature_extraction_utils,
+    pipeline,
+)
+
+###############################################################################
+
+DEFAULT_BASE_MODEL = "superb/wav2vec2-base-superb-sid"
+
+###############################################################################
+
+
+def train(
+    dataset: Union[
+        pd.DataFrame,
+        Dataset,
+        DatasetDict,
+    ],
+    model_name: str = "trained-speakerbox",
+    model_base: str = DEFAULT_BASE_MODEL,
+    max_duration: float = 2.0,
+    batch_size: int = 2,
+) -> Path:
+    """
+    Train a speaker classification model.
+
+    Parameters
+    ----------
+    dataset: Union[pd.DataFrame, Dataset, DatasetDict]
+        The dataset to use for training.
+
+        Should only contain the columns/features: "label" and "audio".
+
+        If provided a pandas DataFrame, this will convert it into a Dataset object
+        and cast the "label" column into ClassLabels and the "audio" column into Audio.
+        If provided a Dataset object (or the converted pandas DataFrame),
+        this will convert it into a DatasetDict with
+        random train, test, and validation splits.
+    model_name: str
+        A name for the model. This will also create a directory with the same name
+        to store the produced model in.
+        Default: "trained-speakerbox"
+    model_base: str
+        The model base to use before fine tuning.
+    max_duration: float
+        The maximum duration to use for each audio clip.
+        Any clips longer than this will be trimmed.
+        Default: 2.0
+    batch_size: int
+        The number of examples to use in a batch during training.
+        Default: 2
+
+    Returns
+    -------
+    model_storage_path: Path
+        The path to the directory where the model is stored.
+    """
+    # Load feature extractor
+    feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(model_base)
+
+    # Convert provided dataset to HF Dataset
+    if isinstance(dataset, pd.DataFrame):
+        dataset = Dataset.from_pandas(dataset, preserve_index=False)
+        dataset = dataset.class_encode_column("label")
+        dataset = dataset.cast_column("audio", Audio(feature_extractor.sampling_rate))
+
+    # Convert into train, test, and validate dict
+    if isinstance(dataset, Dataset):
+        train_and_test = dataset.train_test_split(test_size=0.4)
```

Comment:
What is the size of the train set and is it enough to resist overfitting? Usually, the more complicated a model is the more it is prone to overfitting. Don't know what the recommended size would be given the complexity of this last linear layer of the model though.

Reply:
The splits come out to be train 60% test 20% validation 20%. Summary stats print out:
```
Summary stats for 'train' dataset
n-labels: 8
Avg duration: 1.8705238095238104
Min duration: 0.5
Max duration: 2.0000000000001164
StD duration: 0.3520745452831551
--------------------------------------------------------------------------------
Summary stats for 'test' dataset
n-labels: 8
Avg duration: 1.8874107142857177
Min duration: 0.5
Max duration: 2.0
StD duration: 0.31320475603861686
--------------------------------------------------------------------------------
Summary stats for 'valid' dataset
n-labels: 8
Avg duration: 1.8423928571428592
Min duration: 0.51
Max duration: 2.0
StD duration: 0.38857151712428056
--------------------------------------------------------------------------------
```

Reply:
The general rule of thumb is the training set size should be greater than 10 * the number of effective model parameters (the number of weights of the linear layer of the model that is being fine-tuned). Need to ask a machine learning practitioner if we want to be sure.

Reply:
Not entirely sure about this, but the number of weights would  be around `the dimensions of the output of the feature_extractor * the number of speaker_labels`

If adding more data is too expensive, we can add regularization to nudge the training toward simpler models to resist overfitting. Though I couldn't find that option in the doc.